### PR TITLE
[Permute] Add permute layer forward/backward

### PR DESCRIPTION
**Pull request by zhoonit**
- [Permute] Add permute layer forward/backward

```
This patch implements forward/backward with a following test

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/utils/base_properties.h [hunk 1]
```diff
@@ -20,30 +24,63 @@
...
+using remove_cv_ref_t = std::remove_cv_t<std::remove_reference_t<T>>;
 /**
- * @brief property tag to specialize functions based on this
+ * @brief property info to specialize functions based on this
  *
  * @tparam T property type
  */
-template <typename T> struct prop_tag { using type = typename T::prop_tag; };
+template <typename T> struct prop_info {
+  using prop_type = remove_cv_ref_t<T>;
```
Comment (jijoongmoon) [inline nntrainer/utils/base_properties.h:35]: Add more comments here? what is prop_type, tag_type, data_type with example?

File: nntrainer/layers/permute_layer.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+   */
+  int initialize(Manager &manager) override;
+
+  /**
+   * @copydoc Layer::export_to(Exporter &exporter, ExportMethods method)
+   */
+  void export_to(
+    Exporter &exporter,
```
Comment (jijoongmoon) [inline nntrainer/layers/permute_layer.h:105]: is it for the generate ini?

File: nntrainer/layers/permute_layer.cpp [hunk 3]
```diff
@@ -20,9 +22,57 @@
 
 namespace nntrainer {
 
+bool props::Direction::isValid(const unsigned int &value) const {
```
Comment (jijoongmoon) [inline nntrainer/layers/permute_layer.cpp:25]: is it direction or axis?? which one is better?

File: nntrainer/layers/permute_layer.h [hunk 4]
```diff
@@ -14,11 +14,34 @@
...
 
 #include <array>
+#include <string>
 
 #include <base_properties.h>
 #include <layer_internal.h>
 #include <node_exporter.h>
 namespace nntrainer {
+
+namespace props {
```
Comment (jijoongmoon) [inline nntrainer/layers/permute_layer.h:24]: Are we going to have this props namespace? if it is we need to think more. More code is coming.. so.. maybe we need more namespace like layer.. and optimizer.

File: nntrainer/layers/permute_layer.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/layers/permute_layer.h:105]: This is to generate ini, tflite, get properties.

File: nntrainer/layers/permute_layer.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/layers/permute_layer.h:24]: props namespace are used to explicitly tell it is derived from a property class and can be dealt as other properties are used.
This namespace is used inside `common_properties.h` `base_properties.h` and some custom properties.

```cpp
   props::Name name; // we can explicitly know from the code it is property, and we can expect some common functions like to_string, from_string with this class.
```

vs 
```cpp
  Name name; // we don't know it's a class or a property-ish class.
```
I think it's a policy we can decide. I wonder if we should not use namespace for the props?

File: nntrainer/utils/base_properties.h (hunk 1 above)
Comment (zhoonit) [reply nntrainer/utils/base_properties.h:35]: Updated docs and added test code for clarity :)

File: nntrainer/layers/permute_layer.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/layers/permute_layer.cpp:25]: Thanks for the suggestion. Changed to `PermuteDims` referring from keras properties :)

File: nntrainer/layers/permute_layer.h (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/layers/permute_layer.h:24]: No. I'm saying, If we need, we have to use it other places to maintain consistency.

File: test/unittest/unittest_base_properties.cpp [hunk 5]
```diff
@@ -96,6 +96,19 @@ TEST(BasicProperty, tagCast) {
   }
 }
 
+TEST(BasicProperty, propInfo) {
+  { /**< prop_info test */
+    using prop_type = nntrainer::prop_info<QualityOfBanana>::prop_type;
```
Comment (jijoongmoon) [inline test/unittest/unittest_base_properties.cpp:101]: Let's say differently. for the property input_shape.
then
1. prop_type is input_shape.
2. tag_type is dimension
3. data_type is string 

??

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: test/unittest/unittest_base_properties.cpp (hunk 5 above)
Comment (zhoonit) [reply test/unittest/unittest_base_properties.cpp:101]: Yes, that is current behavior 👍. For `#2`, this is left to incorporate with different property parser. (eg, dimension type for ini but maybe `string` type for tflite.)

File: nntrainer/layers/permute_layer.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/layers/permute_layer.h:24]: `common_properties.h` has props shared across layers.
and each layer can have its own properties. Should we not use namespace for that?

File: nntrainer/layers/permute_layer.cpp [hunk 6]
```diff
@@ -76,10 +76,18 @@ int PermuteLayer::initialize(Manager &manager) {
...
+  auto &input_ = net_input[0]->getVariableRef();
+  auto &hidden_ = net_hidden[0]->getVariableRef();
+
+  input_.transpose(direction_str, hidden_);
 }
 
-void PermuteLayer::calcDerivative() { /** NYI */
+void PermuteLayer::calcDerivative() {
+  auto &input_grad = net_input[0]->getGradientRef();
+  auto &hidden_grad = net_hidden[0]->getVariableRef();
```
Comment (lhs8928) [inline nntrainer/layers/permute_layer.cpp:88]: In the backwarding I think the hidden_grad should be net_hidden[0]->getGradientRef() rather than getVariableRef(), isn't it?

Comment (zhoonit) [reply nntrainer/layers/permute_layer.cpp:88]: Nice catch! this should be getGradientRef(), thanks :)

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/layers/permute_layer.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/layers/permute_layer.cpp:88]: @zhoonit Please update this before merge.

File: nntrainer/layers/permute_layer.cpp [hunk 7]
```diff
@@ -36,7 +36,7 @@ const std::string PermuteLayer::type = "permute";
  * @return const std::string string to return
  */
 static std::string
-buildTrasposeString(const std::array<props::Direction, 3> &arr) {
+buildTrasposeString(const std::array<props::PermuteDims, 3> &arr) {
   std::stringstream ss;
   ss << arr[0].get() << ':' << arr[1].get() << ':' << arr[2].get();
```
Comment (kparichay) [inline nntrainer/layers/permute_layer.cpp:41]: not directly related to this PR, but consider first getting unique values from this array and then applying the transpose operation. Also, the current implementation expects exactly 3 dimensions to be set for permute, but even two dimensions should be enough to use the permute layer.

Comment (zhoonit) [reply nntrainer/layers/permute_layer.cpp:41]: You mean applying array directly? Yes, this should better be supported.

Since we have fixed number of dimension(4), I am concerning that the exact semantics of having two dimension, If this is solid, I think we can apply for two dimension.

Comment (kparichay) [reply nntrainer/layers/permute_layer.cpp:41]: Specifying 3 dimensions for transposing 2 dimensions is an inconvenience. I will approve for this, this can be considered as TODO. Also, there needs to a way to ensure that all the values are unique.

Comment (zhoonit) [reply nntrainer/layers/permute_layer.cpp:41]: > But, there needs to a way to ensure that all the values are unique

Could you check out if this is what you've stated? https://github.com/nnstreamer/nntrainer/blob/6cf6dd2844b4fd6bb49b2728ab06af2432ed60e4/nntrainer/layers/permute_layer.cpp#L47

> Specifying 3 dimensions for transposing 2 dimensions is an inconvenience.

afaik, for other frameworks it is forced that `rank(shape)` `len(direction)` must match, as we have semantics that `rank(shape) == 3`, I think we need stronger assumption or semantics to make it clear.

For example, given ​shape `2:1:4
if 2:1 is given, possible interpretations are
  1. `1 <-> 4`, because h is 1 when starting from channel
  1. invalid, because height 1 means there is height does not count.
  1. invalid, because `len(direction) == 2` must start from zero

if 0:1 is given, possible interpretations are
  1. `2 <-> 1` starting from channel
  1. `1 <-> 4` starting from height (but invalid because channel is > 2)
  1.  `2 <-> 4` starting from channel but height is 1 so it is not counted.

If we have solid interpretation considering we have fixed number of dimensions I think we could support num 2.

Comment (kparichay) [reply nntrainer/layers/permute_layer.cpp:41]: > Could you check out if this is what you've stated?

Thanks for the link. I missed this. For later, you may consider moving such constraints to setProperty().

Agreed on the part that 2 dimensions might lead to confusion. 
However, this raises question (outside the scope of this PR), dimension `0:1:2` maps to `c:h:w` and in some cases, `0:1:2:3` maps to `b:c:h:w`. Let's make it consistent.

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/layers/permute_layer.cpp (hunk 7 above)
Comment (zhoonit) [reply nntrainer/layers/permute_layer.cpp:41]: >  this raises question

Yes, for now, we have firm semantics it is always `c:h:w` and current behavior (ignoring batch dimension) coherent with keras `PermuteLayer`. but `tf.transpose` will start from batch. 


Please throw me ideas if any.

My current thinking is transpose in the near future, so we might be needing two semantics (with different name I guess :p)

Or we can consider it as one semantic, i guess, if we never allow putting batch dimension inside *.ini or property.

Thank you for raising the issue :)

Comment (kparichay) [reply nntrainer/layers/permute_layer.cpp:41]: https://www.tensorflow.org/api_docs/python/tf/keras/layers/Permute
tf.keras.layers.permute starts dimensions with 1 (I think they left 0 for batch). 
Let's follow these semantics to avoid confusion.

Comment (zhoonit) [reply nntrainer/layers/permute_layer.cpp]: Thanks, made a change to accept 1~3
